### PR TITLE
Async writes and session lock fixes

### LINKED ISSUE #3133: Improve locking and async for tornado server side
There are several entangled problems, I have partial code already to fix this. We should merge https://github.com/bokeh/bokeh/pull/3119 as-is first and fix that along with the rest.

 * periodic and timeout callbacks need to `session._lock.acquire()`
     See https://github.com/bokeh/bokeh/blob/2a3ad528279e3b329857d712b61fcca536a59cf8/bokeh/server/session.py#L78-L80
 * when the periodic callback is fixed to acquire the lock, it becomes async, which means it becomes pesky to keep them from "piling up"; https://groups.google.com/d/msg/python-tornado/yprQnT61zps/wfR6zNycGzsJ shows a possible solution
 * we don't `yield` on writes; meaning these `write_message` are done as fire-and-forget I believe (we ignore the returned future): https://github.com/bokeh/bokeh/blob/2a3ad528279e3b329857d712b61fcca536a59cf8/bokeh/server/protocol/message.py#L171-L180
 * if we did `yield` on `write_message`, we don't have any locking to prevent other writes during the yield (might happen due to a periodic callback for example). to fix this we might want a `connection._lock`
 * making writes async means that in `session._document_changed` we need a way to move the patch-doc write to an async context, the way I started to do this is to collect `patch-doc` and make them part of the "work" returned by the handler to here (work can now be a list): https://github.com/bokeh/bokeh/blob/2a3ad528279e3b329857d712b61fcca536a59cf8/bokeh/server/views/ws.py#L149-L171
  * if we wanted to we could hold a `connection._lock` across an entire handler and processing its work, which could keep semantics more predictable (avoids interleaving events that result from distinct callbacks)

Anyway it may be clearer as a patch, which I'm working on.

### PR DESCRIPTION
**Pull request by havocp**
This is supposed to close #3133 

TLDR fixes here are:
 * avoid backlog by waiting for writes and periodic callbacks to finish before queueing up more
 * take the session._lock when we run periodic and timeout callbacks

The avoid-write-backlog thing only "really" works with Tornado 4.3, which isn't in conda right now - so I'm testing here with compat shims that simulate 4.3 without actually waiting on the writes.
I guess that's better than hard-requiring 4.3, though with 4.3 things should work better on a loaded server.

### CODE REVIEW

**Comment by bryevdv** [comment]
OK not sure when 4.3 will be available as a conda pkg we can review the dependency requirement once it is.
